plugins/catalog-react/src/hooks/useEntityPermission.ts

```diff
@@ -19,12 +19,19 @@ import { Permission } from '@backstage/plugin-permission-common';
 import { usePermission } from '@backstage/plugin-permission-react';
 import { useEntity } from './useEntity';
 
+const cache: Record<string, boolean> = {};
```

Comment:
This is likely going to be problematic one way or the other, global state is almost universally evil :) Firstly, it's shared everywhere always unconditionally, which means that it'll also leak across tests and stuff. Secondly, there's absolutely no eviction mechanism; there should at the very least be a smallish TTL on entries (so a lightweight small third party cache lib of some kind may be interesting, rather than making it ourselves).

Honestly I'd prefer for this to not be part of the hook logic at all. It's better if the underlying client invoked by `usePermission` dealt with this transparently. Probably not even as the root client, but as a wrapper.

Pseudocode for the lowest levels:

```ts
const client = new CachingPassthroughClient({
  inner: new ActualClient(...),
  ttl: Duration.fromObject({ minutes: 1 }),
})
```

Reply:
It kinda needs to be at the hook level, since the client isn't able to say "wait I'm thinking, but for now assume allow", and later go "no you can't actually do that". Makes sense to put this further up the chain in `usePermission` though?

Definitely share the concerns about global state though, so let's handle this with care x)

Reply:
what about storing these things in a context instead? I'm imagining a lot of components want to react and refetch things when the state changes. Is there any generic context in backstage that can be used for this purpose?
We might lose some of the React's capabilities that context are giving us for free If we cache this data inside the client

Reply:
I'm thinking that every component asks what it needs to ask on mount, and eventually that promise resolves. Either almost instantly, or some time later. No state updates to deal with after that, unless we want to make it possible to for example react (without a page reload) to a server-side permissions revocation or whatnot.

Reply:
A possible case might be a user who log out and login with another account, staying on the same component. It would be nice to react on this change automatically

Reply:
We talked internally a bit though.

If we want to be able to also support stale-while-revalidate, we could either make the client return an Observable, or add a (synchronous) hint getter to the client interface that can be used to get instant feedback about allow/reject/allow-but-check-async-too/reject-but-check-async-too/do-not-know-check-async-instead, and then keeping the full async one as is (except it writes outcomes to its internal cache)

Reply:
Ah but logout is a bit of a different beast isn't it. That means that busting the cache isn't even enough, it means you have to really _actively_ reconsider every single auth request because the circumstances have changed. Still, yeah, an observable would support that too, but it would effectively make permission requests a "live query" rather than a point consideration. Interesting.

Reply:
doesn't context give us all of these capabilities already? Here we want to re-render all the connected components if something changes. Sure, at the end the outcome will be the same, but maintaining a lot of long living observables (with usually a lot of them subscribing to the same data) while manually making them update some react state looks more complicated.

Reply:
Another route we discussed is to add something like the following to the frontend API:

```ts
  getTenativeAuthorization(request: AuthorizeQuery): AuthorizeDecision;
```

Reply:
I should've known that a caching PR was never going to be this simple... "hardest problem in CS" and all that. 😄 

It's a good point about SWR - since that's basically what we want to achieve, if we're ok with bringing in libraries for it, I can wire one up.

I do think logout is a rare enough case that we can probably tackle a solution to that separately if it becomes necessary.

Reply:
Hmm, so instead of adding this as a functionality of the `usePermission` hook, it would be a separate feature baked into the frontend API itself?

Personally I am leaning towards baking it into the `usePermission` hook because it means consumers don't have to think about whether they need to call a different method first.

Reply:
If you do it in the hook and need it to be caching across components, you still need an external state store. So then you'd be back at either putting it in a context (arguably annoying to have to change the jsx shape and feature set just to get the feature), or in "js land" outside of the react tree which ends up being observables or callback handlers etc

Reply:
I think wrapping the api call inside of the `usePermission` hook with [`useSWR`](https://swr.vercel.app/) would neatly abstract away all the work that's happening outside of the React tree. Would a separate method on the frontend API still be preferred to something like this?

Reply:
Just pushed up an implementation using `useSWR`, let me know what you think!